server/findproviders.go

```diff
@@ -1,98 +1,141 @@
+	rch := make(chan *proto.DelegatedRouting_FindProviders_AsyncResult)
+	go func() {
+		defer close(rch)
+		pcids := parseCidsFromFindProvidersRequest(req)
+		for _, c := range pcids {
+			ch, err := drs.service.FindProviders(c)
+			if err != nil {
+				logger.Errorf("find providers function rejected request (%w)", err)
+				continue
 			}
 			for x := range ch {
+				var resp *proto.DelegatedRouting_FindProviders_AsyncResult
 				if x.Err != nil {
-					log.Errorf("get p2p provider returned error (%v)", x.Err)
-					continue
-				}
-				resp := GenerateGetP2PProvideResponse(x.AddrInfo)
-				env := &parser.Envelope{
-					Tag:     parser.MethodGetP2PProvide,
-					Payload: resp,
-				}
-				enc, err := json.Marshal(env)
-				if err != nil {
-					continue
+					logger.Infof("find providers function returned error (%w)", x.Err)
+					resp = &proto.DelegatedRouting_FindProviders_AsyncResult{Err: x.Err}
+				} else {
+					resp = buildFindProvidersResponse(c, x.AddrInfo)
 				}
-				writer.Write(enc)
+				rch <- resp
 			}
-		default:
-			log.Errorf("unknown method (%v)", env.Tag)
-			writer.WriteHeader(404)
 		}
-	}
+	}()
+	return rch, nil
+}
+
+func parseCidsFromFindProvidersRequest(req *proto.FindProvidersRequest) []cid.Cid {
+	return []cid.Cid{cid.Cid(req.Key)}
 }
 
-// ParseGetP2PProvideRequest parses a GetP2PProvideRequest and returns the included bytes key in the form of a cid.
-func ParseGetP2PProvideRequest(req *parser.GetP2PProvideRequest) (cid.Cid, error) {
-	mhBytes, err := parser.FromDJSpecialBytes(req.Key)
-	if err != nil {
-		return cid.Undef, err
+func buildFindProvidersResponse(key cid.Cid, addrInfo []peer.AddrInfo) *proto.DelegatedRouting_FindProviders_AsyncResult {
+	provs := make(proto.ProvidersList, len(addrInfo))
+	bitswapProto := proto.TransferProtocol{Bitswap: &proto.BitswapProtocol{}}
+	for i, addrInfo := range addrInfo {
+		provs[i] = proto.Provider{
+			ProviderNode:  proto.Node{Peer: buildPeerFromAddrInfo(addrInfo)},
+			ProviderProto: proto.TransferProtocolList{bitswapProto},
+		}
 	}
-	parsedCid := cid.NewCidV1(cid.Raw, mhBytes)
-	if err != nil {
-		return cid.Undef, err
+	return &proto.DelegatedRouting_FindProviders_AsyncResult{
+		Resp: &proto.FindProvidersResponse{Providers: provs},
 	}
-	return parsedCid, nil
 }
 
-func GenerateGetP2PProvideResponse(infos []peer.AddrInfo) *parser.GetP2PProvideResponse {
-	resp := &parser.GetP2PProvideResponse{}
-	for _, info := range infos {
-		for _, addr := range info.Addrs {
-			peerAddr := addr.Encapsulate(multiaddr.StringCast("/p2p/" + info.ID.String()))
-			resp.Peers = append(resp.Peers, parser.ToDJSpecialBytes(peerAddr.Bytes()))
+func buildPeerFromAddrInfo(addrInfo peer.AddrInfo) *proto.Peer {
+	pm := make([]values.Bytes, len(addrInfo.Addrs))
+	for i, addr := range addrInfo.Addrs {
+		// XXX: Adin, please, verify this is as intended.
```

Comment:
@aschmahmann Please, verify this is as intended. PeerAddrs are created from whatever the server-implementing user passes. The code rewrites the multiaddresses to make sure they conform to the spec.